policy/resolved_policy_builder.go

```diff
@@ -0,0 +1,1132 @@
+			if action != explorer.Action_UNSPECIFIED && action != explorer.Action_MODIFY {
+				actions[pRef.Mrn] = action
+			}
+
+			if pRef.Impact != nil {
+				impacts[pRef.Mrn] = pRef.Impact
+			}
+			scoringSystem := pRef.ScoringSystem
+
+			if scoringSystem != explorer.ScoringSystem_SCORING_UNSPECIFIED {
+				scoringSystems[pRef.Mrn] = pRef.ScoringSystem
+			} else {
+				if p, ok := b.bundleMap.Policies[pRef.Mrn]; ok {
+					scoringSystems[pRef.Mrn] = p.ScoringSystem
+				}
+			}
+		}
+
+		// We always want to select the worst impact that we find
+		getWorstImpact := func(impact1 *explorer.Impact, impact2 *explorer.Impact) *explorer.Impact {
+			if impact1 == nil {
+				return impact2
+			}
+			if impact2 == nil {
+				return impact1
+			}
+
+			if impact1.Value.GetValue() > impact2.Value.GetValue() {
+				return impact1
+			}
+			return impact2
+		}
+
+		for _, c := range g.Checks {
+			impact := c.Impact
+			action := normalizeAction(g.Type, c.Action, impact)
+			if action == explorer.Action_IGNORE {
+				// Since we traversed the children first, we overwrite the impact with ignore
+				// if the action is ignore. We do not merge in this case
+				impact = &explorer.Impact{
+					Scoring: explorer.ScoringSystem_IGNORE_SCORE,
+				}
+			} else {
+				if qBundle, ok := b.bundleMap.Queries[c.Mrn]; ok {
+					// Check the impact defined on the query
+					impact = getWorstImpact(impact, qBundle.Impact)
+				}
+
+				impact = getWorstImpact(impact, impacts[c.Mrn])
+			}
+
+			if action != explorer.Action_UNSPECIFIED && action != explorer.Action_MODIFY {
+				actions[c.Mrn] = action
+			}
+
+			if impact != nil {
+				impacts[c.Mrn] = impact
+			}
+		}
+
+		for _, q := range g.Queries {
+			if q.Action != explorer.Action_UNSPECIFIED {
+				action := normalizeAction(g.Type, q.Action, q.Impact)
+				if action != explorer.Action_UNSPECIFIED && action != explorer.Action_MODIFY {
+					actions[q.Mrn] = action
+				}
+			}
+		}
+	}
+
+	for _, r := range policy.RiskFactors {
+		if r.Magnitude != nil {
+			riskMagnitudes[r.Mrn] = r.Magnitude
+		}
+
+		if r.Action != explorer.Action_UNSPECIFIED && r.Action != explorer.Action_MODIFY {
+			actions[r.Mrn] = r.Action
+		}
+	}
+}
+
+func canRun(action explorer.Action) bool {
+	return !(action == explorer.Action_DEACTIVATE || action == explorer.Action_OUT_OF_SCOPE)
+}
+
+type group interface {
+	GetReviewStatus() ReviewStatus
+	GetEndDate() int64
+}
+
+type groupWithFilters interface {
+	GetFilters() *explorer.Filters
+}
+
+// isGroupMatching checks if the policy group is matching. A policy group is matching if it is not rejected,
+// and it is not expired. If it has filters, it must have at least one filter that matches the asset filters
+func (b *resolvedPolicyBuilder) isGroupMatching(group group) bool {
+	if group.GetReviewStatus() == ReviewStatus_REJECTED {
+		return false
+	}
+
+	if group.GetEndDate() != 0 {
+		// TODO: we also need to check if the group is accepted or rejected
```

Comment:
Is this something that needs to happen now, before this is merged?

Reply:
comment is wrong, it happens in the lines above. will remove